File: Others/subarraySumEqualsK/optimize.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

// [2,4,1,5,3,6]   k = 6
//        ^

// memo[tmp], # subarrys whose sum is equal to tmp
int subarraySum(std::vector<int> &nums, int k) {
    std::unordered_map<int, int> memo;
    memo[0] = 1;
    int res = 0, cur_sum = 0;
    for (auto &num:nums) {
        cur_sum += num;
        res += memo[cur_sum - k];
        memo[cur_sum] += 1;
    } 
    return res;
}

int main() {
    std::vector<int> nums;
    nums.push_back(2);
    nums.push_back(4);
    nums.push_back(1);
    nums.push_back(5);
    nums.push_back(3);
    nums.push_back(6);
    
    int k = 6;
    int res = subarraySum(nums, k);
    std::cout << "res: " << res << "\n";
}

File: Others/subarraySumEqualsK/optimize.java
import java.util.Map;
import java.util.HashMap;

public class optimize {
    public static int subarraySum(int[] nums, int k) {
        int res = 0, tmp = 0;
        Map<Integer, Integer> memo = new HashMap<>();
        memo.put(0, 1);
        for (int i = 0; i < nums.length; ++i) {
            tmp += nums[i];
            if (memo.containsKey(tmp - k))
                res += memo.get(tmp - k);
            memp[tmp] = memo.getDefault(tmp, 0) + 1; 
        }
        return res;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1,1,1};
        int k = 2;
        int res = subarraySum(nums, k);
        System.out.println(res);
    }
}
